ISPN-14119 Filter null from evicted entities for distributed queries

fax4ever opened the pull request:
https://issues.redhat.com/browse/ISPN-14119

wburns reviewed (changes requested):
From the sound of the bug, my guess would be is the entry is in the index, the user runs the query which finds the reference id, and concurrently the entry is removed due to expiration and thus when it tries to read the referenced key it is no longer present.

I would suggest as a guaranteed way to reproduce this issue I would recommend using Mocks.blockingMock replacing the component that does the actual object lookup (if possible). This allows you to fire the query in one thread and wait until it gets to where it is looking up the entries and block the operation. Then you can increment the time service and expire the entry. Then let the query continue and it should face the issue every time you run it without having to do loops or sleeps.

fax4ever commented:
Thanks for review @wburns!

> From the sound of the bug, my guess would be is the entry is in the index, the user runs the query which finds the reference id, and concurrently the entry is removed due to expiration and thus when it tries to read the referenced key it is no longer present.

exactly!

I introduced the controlled time service and ran the query before and after the expiration.
I'm going to find if it possible to use the `Mocks.blockingMock` on entity loading phase.
I was asking myself if we should do that for all tests?

wburns commented:
> I introduced the controlled time service and ran the query before and after the expiration. I'm going to find if it possible to use the `Mocks.blockingMock` on entity loading phase. I was asking myself if we should do that for all tests?

Not sure what you mean by all tests. But I would say if you are using a test that has expiration, you should definitely use a controlled time service, so you don't have the tests take too long and you control the behavior at a much granular level and should be very easily repeatable.

Regarding blockingMock, that is only needed if there is a step in between. Does the test as you have it reproduce the issue every time it is invoked? Or does it fail rarely/intermittently? If so then we will want to inject the blockingMock to be sure to reproduce it easily enough.

fax4ever commented:
> Does the test as you have it reproduce the issue every time it is invoked? Or does it fail rarely/intermittently? If so then we will want to inject the blockingMock to be sure to reproduce it easily enough.

I don't think so. Thus I think we can avoid using the blockingMock IIUC.

Thanks again @wburns 
Is there anything else we can improve here?

wburns commented:
> Thanks again @wburns Is there anything else we can improve here?

Seems fine, I am just wondering why we don't have the query module test anymore. Seems a bit much with the rest test to test this specific bug. I am fine having both, but can we resurrect the old test you had as well?

fax4ever commented:
Thank @wburns.
The exception is raised only with the remote queries. In case of an embedded query a list containing some `null` value is returned. I added a test case for it. I think that it is good to have them both (embedded and remote). Updated the contribution.

wburns commented:
Integrated into main, thanks @fax4ever !

fax4ever commented:
thank you Will!
